### DISCUSSION TITLE
Detecting what note has been pressed on a MIDI keyboard.

### DISCUSSION THREAD

**Discussion by Kinokoda** [FortySevenEffects/arduino_midi_library#277] [Q&A] [answered]
I have been trying to find some way to detect what note was pressed but I don't know how to make an if statement for that.

My goal is to find what note has been sent to the Arduino, and then depending on what note it was send an analog voltage out with analogWrite();.

Any ideas?

**Comment by franky47** [comment, marked as answer]
Check out the [Callbacks](https://github.com/FortySevenEffects/arduino_midi_library/blob/master/examples/Callbacks/Callbacks.ino) example, it does just that.

The `handleNoteOn` function will be called when the Arduino receives a NoteOn, and the `pitch` parameter will be the note number (between 0 and 127, check out [this table](https://www.inspiredacoustics.com/en/MIDI_note_numbers_and_center_frequencies) to map it to human-friendly note numbers).

https://github.com/FortySevenEffects/arduino_midi_library/blob/2d64cc3c2ff85bbee654a7054e36c59694d8d8e4/examples/Callbacks/Callbacks.ino#L12-L19

**Reply by franky47** [reply]
For example, to make a naive V/oct converter (assuming a 5V maximum output):

```c++
static const byte lowNote    = 36 // C2;
static const byte maxVoltage = 5; // volts
static const byte minPitch   = lowNote;
static const byte maxPitch   = lowNote + maxVoltage * 12;

void handleNoteOn(byte channel, byte pitch, byte velocity)
{
  const byte clampedPitch = constaint(pitch, minPitch, maxPitch);
  const byte analogOutput = map(clampedPitch, minPitch, maxPitch, 0, 255);
  analogWrite(analogPin, analogOutput);
}
```

**Reply by Kinokoda** [reply]
> 

Where would i need to write this code for it to be used?

**Reply by franky47** [reply]
You can copy the Callbacks example and add that code (replace the `handleNoteOn` function). You'll also have to define an analog pin to use:
```c++
static const int analogPin = 3;
```
Check out which pins you can use:
https://www.arduino.cc/reference/en/language/functions/analog-io/analogwrite/

**Reply by Kinokoda** [reply]
is there a specific command for checking what note has been pressed?

**Reply by Kinokoda** [reply]
I think i slapped it together the right way.

#include <MIDI.h>

MIDI_CREATE_DEFAULT_INSTANCE();

// -----------------------------------------------------------------------------

// This function will be automatically called when a NoteOn is received.
// It must be a void-returning function with the correct parameters,
// see documentation here:
// https://github.com/FortySevenEffects/arduino_midi_library/wiki/Using-Callbacks

static const int analogPin = 5;
static const byte lowNote    = 36 // C2;
static const byte maxVoltage = 5; // volts
static const byte minPitch   = lowNote;
static const byte maxPitch   = lowNote + maxVoltage * 12;

void handleNoteOn(byte channel, byte pitch, byte velocity)
{
  const byte clampedPitch = constaint(pitch, minPitch, maxPitch);
  const byte analogOutput = map(clampedPitch, minPitch, maxPitch, 0, 255);
  analogWrite(analogPin, analogOutput);
}

void handleNoteOff(byte channel, byte pitch, byte velocity)
{
    // Do something when the note is released.
    // Note that NoteOn messages with 0 velocity are interpreted as NoteOffs.
}

// -----------------------------------------------------------------------------

void setup()
{
    // Connect the handleNoteOn function to the library,
    // so it is called upon reception of a NoteOn.
    MIDI.setHandleNoteOn(handleNoteOn);  // Put only the name of the function

    // Do the same for NoteOffs
    MIDI.setHandleNoteOff(handleNoteOff);

    // Initiate MIDI communications, listen to all channels
    MIDI.begin(MIDI_CHANNEL_OMNI);
}

void loop()
{
    // Call MIDI.read the fastest you can for real-time performance.
    MIDI.read();

    // There is no need to check if there are messages incoming
    // if they are bound to a Callback function.
    // The attached method will be called automatically
    // when the corresponding message has been received.
}

**Reply by Kinokoda** [reply]
https://docs.google.com/spreadsheets/d/1EnHLPhJhiaZ5cvquiGJeRmWuxxn5R8PR9w2hlqWHvGk/edit?usp=sharing

These are the notes im trying to react to

**Reply by franky47** [reply]
I see. Using [that table](https://www.inspiredacoustics.com/en/MIDI_note_numbers_and_center_frequencies), you can get the MIDI note numbers for each of those notes, then you can use a switch/case statement (don't forget the `break` statements):

```c++
void handleNoteOn(byte channel, byte pitch, byte velocity)
{
  switch (pitch)
  {
    case 36: // C2
      analogWrite(analogPin, 121);
      break;
    case 37: // C#2
      analogWrite(analogPin, 119);
      break;
    // etc...
  }
}
```

**Reply by Kinokoda** [reply]
```c++
#include <MIDI.h>

MIDI_CREATE_DEFAULT_INSTANCE();

// -----------------------------------------------------------------------------

// This function will be automatically called when a NoteOn is received.
// It must be a void-returning function with the correct parameters,
// see documentation here:
// https://github.com/FortySevenEffects/arduino_midi_library/wiki/Using-Callbacks
static const int analogPin = 5;

void handleNoteOn(byte channel, byte pitch, byte velocity)
{
  switch (pitch)
  {
    case 24: // C2
      analogWrite(analogPin, 121);
      break;
    case 25: // C#2
      analogWrite(analogPin, 119);
      break;
    case 26: // D2
      analogWrite(analogPin, 116);
      break;
    case 27: // Eb2
      analogWrite(analogPin, 113);
      break;
    case 28: // E2
      analogWrite(analogPin, 110);
      break;
    case 29: // F2
      analogWrite(analogPin, 105);
      break;
    case 30: // F#2
      analogWrite(analogPin, 102);
      break;
    case 31: // G2
      analogWrite(analogPin, 97);
      break;
    case 32: // G#2
      analogWrite(analogPin, 94);
      break;
    case 33: // A2
      analogWrite(analogPin, 89);
      break;
    case 34: // A#2
      analogWrite(analogPin, 84);
      break;
    case 35: // B2
      analogWrite(analogPin, 79);
      break;
    case 36: // C3
      analogWrite(analogPin, 73);
      break;
  }
}

void handleNoteOff(byte channel, byte pitch, byte velocity)
{
    // Do something when the note is released.
    // Note that NoteOn messages with 0 velocity are interpreted as NoteOffs.
}

// -----------------------------------------------------------------------------

void setup()
{
    // Connect the handleNoteOn function to the library,
    // so it is called upon reception of a NoteOn.
    MIDI.setHandleNoteOn(handleNoteOn);  // Put only the name of the function

    // Do the same for NoteOffs
    MIDI.setHandleNoteOff(handleNoteOff);

    // Initiate MIDI communications, listen to all channels
    MIDI.begin(MIDI_CHANNEL_OMNI);
}

void loop()
{
    // Call MIDI.read the fastest you can for real-time performance.
    MIDI.read();

    // There is no need to check if there are messages incoming
    // if they are bound to a Callback function.
    // The attached method will be called automatically
    // when the corresponding message has been received.
}
```

Is this correct?

**Reply by franky47** [reply]
Yes, that looks good!

**Reply by Kinokoda** [reply]
Oh also, will the note stop playing automatically?

**Reply by franky47** [reply]
No, that would depend on the thing you are controlling with that voltage.

Usually on analog synths, there are envelopes (ADSR) that are in charge of opening or closing a voltage controlled amplifier (VCA), which lets the sound pass when the envelope is active (note on) and mute the output when it's inactive (note off). The controlling of envelopes is done with a Gate signal, which you can generate via a digital pin (HIGH = active, LOW = inactive, though some synths use the opposite convention).

Check out the [SimpleSynth](https://github.com/FortySevenEffects/arduino_midi_library/blob/master/examples/SimpleSynth/SimpleSynth.ino) example, it has the concept of a Gate pin, plus a way to handle a sequence of notes in memory so that you can press multiple keys, the latest one will be the current pitch and only when there are no more keys shall the envelope be released.

**Reply by Kinokoda** [reply]
Oh okay
Dont worry, I have designed a vca and a adsr envelope generator, thanks for telling me though. Ill close the q&a and again thx for he help.

**Reply by Kinokoda** [reply]
hey @franky47, about the gate control.
Can I do it just by setting a pin high with digitalWrite and use that as a gate output?

Im talking about gate control

and will the analogWrite override itself?

**Reply by franky47** [reply]
If your envelope is OK with the Vcc voltage of your Arduino (5V or 3.3V depending on which board you're using), then it should be enough. Otherwise you'd need a piece of circuitry to boost the level (Schmitt trigger, which usually invert their input so you'd have to digitalWrite in the opposite polarity).

Not sure what you mean by "override itself"? Whenever there is a note on, the pitch will be updated to reflect it, but if you were to press a C3, keep it pressed and make a chord with a E3, when releasing the E3 it wouldn't update back to C3, for that you'll need to use the note list code in SimpleSynth.

**Reply by Kinokoda** [reply]
okay then thx
